File: lab/javascript/tecnicas/lista_encadeada.js
class No{
    constructor(valor){
        this.valor = valor;
        this.proximo = null;
    }
}


class ListaEncadeada{
    constructor(){
        this.head = null;  // Começo da lista
    }

    inserir(valor) {
        const novo_no = new No(valor);

        // Se a lista estiver vazia
        if (this.head === null){
            this.head = novo_no;
            return;
        }

        // Se já houver elementos, percorre até o final
        let atual = this.head;
        while (atual.proximo !== null) {
            atual = atual.proximo;
        }

        // Adiciona o novo nó no final
        atual.proximo = novo_no;
    }

    imprimir(){
        let atual = this.head;
        let resultado = '';
        while (atual !== null) {
        resultado += atual.valor + ' → ';
        atual = atual.proximo;
        }
        console.log(resultado + 'null');
    }
}


const lista = new ListaEncadeada();
lista.inserir(10);
lista.inserir(20);
lista.inserir(30);
lista.imprimir();

File: lab/python/tecnicas/lista_encadeada.py
class No:
    def __init__(self, valor) -> None:
        self.valor = valor
        self.proximo = None


class ListaEncadeada:
    def __init__(self) -> None:
        self.head = None  # Começo da lista

    def inserir(self, valor):
        novo_no = No(valor)

        # Se a lista estiver vazia
        if self.head is None:
            self.head = novo_no
            return

        # Se já houver elementos, percorre até o final
        atual = self.head
        while atual.proximo is not None:
            atual = atual.proximo

        # Adiciona o novo nó no final
        atual.proximo = novo_no

    def imprimir(self):
        atual = self.head
        while atual is not None:
            print(atual.valor, end=" → ")
            atual = atual.proximo
        print("None")


if __name__ == "__main__":
    lista = ListaEncadeada()
    lista.inserir(10)
    lista.inserir(20)
    lista.inserir(30)
    lista.imprimir()
